Prevent interface conversion panic

Hi!

I am trying to generate some protobuf using https://github.com/google/gnostic-grpc, which depends on protoreflect.

The openapi spec I am using is unfortunately generating some conflicting message/enum names, at least based on the error message I am receiving:

panic: interface conversion: desc.Descriptor is *desc.MessageDescriptor, not *desc.EnumDescriptor

It would be very helpful to have a proper error message, rather than a panic, when the interface conversion fails here:

https://github.com/jhump/protoreflect/blob/05026f325d15a06428b5dbb7e27cbb92b0be249a/desc/descriptor.go#L558

--- a/desc/descriptor.go
+++ b/desc/descriptor.go
@@ -555,25 +555,37 @@ func (fd *FieldDescriptor) resolve(path []int32, scopes []scope) error {
 	}
 	fd.sourceInfoPath = append([]int32(nil), path...) // defensive copy
 	if fd.proto.GetType() == dpb.FieldDescriptorProto_TYPE_ENUM {
-		if desc, err := resolve(fd.file, fd.proto.GetTypeName(), scopes); err != nil {
+		desc, err := resolve(fd.file, fd.proto.GetTypeName(), scopes)
+		if err != nil {
 			return err
-		} else {
-			fd.enumType = desc.(*EnumDescriptor)
 		}
+		enumType, ok := desc.(*EnumDescriptor)
+		if !ok {
+			return fmt.Errorf("could not convert desc %s to EnumDescriptor. TypeName: %s", desc, fd.proto.GetTypeName())
+		}
+		fd.enumType = enumType
 	}
 	if fd.proto.GetType() == dpb.FieldDescriptorProto_TYPE_MESSAGE || fd.proto.GetType() == dpb.FieldDescriptorProto_TYPE_GROUP {
-		if desc, err := resolve(fd.file, fd.proto.GetTypeName(), scopes); err != nil {
+		desc, err := resolve(fd.file, fd.proto.GetTypeName(), scopes)
+		if err != nil {
 			return err
-		} else {
-			fd.msgType = desc.(*MessageDescriptor)
 		}
+		msgType, ok := desc.(*MessageDescriptor)
+		if !ok {
+			return fmt.Errorf("could not convert desc %s to MessageDescriptor. TypeName: %s", desc, fd.proto.GetTypeName())
+		}
+		fd.msgType = msgType
 	}
 	if fd.proto.GetExtendee() != "" {
-		if desc, err := resolve(fd.file, fd.proto.GetExtendee(), scopes); err != nil {
+		desc, err := resolve(fd.file, fd.proto.GetExtendee(), scopes)
+		if err != nil {
 			return err
-		} else {
-			fd.owner = desc.(*MessageDescriptor)
 		}
+		msgType, ok := desc.(*MessageDescriptor)
+		if !ok {
+			return fmt.Errorf("could not convert desc %s to MessageDescriptor. TypeName: %s", desc, fd.proto.GetTypeName())
+		}
+		fd.owner = msgType
 	}
 	fd.file.registerField(fd)
 	fd.isMap = fd.proto.GetLabel() == dpb.FieldDescriptorProto_LABEL_REPEATED &&

--- a/desc/descriptor.go
+++ b/desc/descriptor.go
@@ -549,6 +549,32 @@ func createFieldDescriptor(fd *FileDescriptor, parent Descriptor, enclosing stri
 	return ret, fldName
 }
 
+func descriptorType(d Descriptor) string {
+	switch d := d.(type) {
+	case *FileDescriptor:
+		return "a file"
+	case *MessageDescriptor:
+		return "a message"
+	case *FieldDescriptor:
+		if d.IsExtension() {
+			return "an extension"
+		}
+		return "a field"
+	case *OneOfDescriptor:
+		return "a oneof"
+	case *EnumDescriptor:
+		return "an enum"
+	case *EnumValueDescriptor:
+		return "an enum value"
+	case *ServiceDescriptor:
+		return "a service"
+	case *MethodDescriptor:
+		return "a method"
+	default:
+		return fmt.Sprintf("a %T", d)
+	}
+}
+
 func (fd *FieldDescriptor) resolve(path []int32, scopes []scope) error {
 	if fd.proto.OneofIndex != nil && fd.oneOf == nil {
 		return fmt.Errorf("could not link field %s to one-of index %d", fd.fqn, *fd.proto.OneofIndex)
@@ -561,7 +587,7 @@ func (fd *FieldDescriptor) resolve(path []int32, scopes []scope) error {
 		}
 		enumType, ok := desc.(*EnumDescriptor)
 		if !ok {
-			return fmt.Errorf("could not convert desc %s to EnumDescriptor. TypeName: %s", desc, fd.proto.GetTypeName())
+			return fmt.Errorf("field %v indicates a type of enum, but references %q which is %s", fd.fqn, fd.proto.GetTypeName(), descriptorType(desc))
 		}
 		fd.enumType = enumType
 	}
@@ -572,7 +598,7 @@ func (fd *FieldDescriptor) resolve(path []int32, scopes []scope) error {
 		}
 		msgType, ok := desc.(*MessageDescriptor)
 		if !ok {
-			return fmt.Errorf("could not convert desc %s to MessageDescriptor. TypeName: %s", desc, fd.proto.GetTypeName())
+			return fmt.Errorf("field %v indicates a type of message, but references %q which is %s", fd.fqn, fd.proto.GetTypeName(), descriptorType(desc))
 		}
 		fd.msgType = msgType
 	}
@@ -583,7 +609,7 @@ func (fd *FieldDescriptor) resolve(path []int32, scopes []scope) error {
 		}
 		msgType, ok := desc.(*MessageDescriptor)
 		if !ok {
-			return fmt.Errorf("could not convert desc %s to MessageDescriptor. TypeName: %s", desc, fd.proto.GetTypeName())
+			return fmt.Errorf("field %v extends %q which should be a message but is %s", fd.fqn, fd.proto.GetExtendee(), descriptorType(desc))
 		}
 		fd.owner = msgType
 	}
@@ -1131,6 +1157,7 @@ func (sv sortedValues) Less(i, j int) bool {
 
 func (sv sortedValues) Swap(i, j int) {
 	sv[i], sv[j] = sv[j], sv[i]
+
 }
 
 func (ed *EnumDescriptor) resolve(path []int32) {
@@ -1453,12 +1480,20 @@ func (md *MethodDescriptor) resolve(path []int32, scopes []scope) error {
 	if desc, err := resolve(md.file, md.proto.GetInputType(), scopes); err != nil {
 		return err
 	} else {
-		md.inType = desc.(*MessageDescriptor)
+		msgType, ok := desc.(*MessageDescriptor)
+		if !ok {
+			return fmt.Errorf("method %v has request type %q which should be a message but is %s", md.fqn, md.proto.GetInputType(), descriptorType(desc))
+		}
+		md.inType = msgType
 	}
 	if desc, err := resolve(md.file, md.proto.GetOutputType(), scopes); err != nil {
 		return err
 	} else {
-		md.outType = desc.(*MessageDescriptor)
+		msgType, ok := desc.(*MessageDescriptor)
+		if !ok {
+			return fmt.Errorf("method %v has response type %q which should be a message but is %s", md.fqn, md.proto.GetOutputType(), descriptorType(desc))
+		}
+		md.outType = msgType
 	}
 	return nil
 }

